src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs

```diff
@@ -0,0 +1,2164 @@
+                lock (_lockObject)
+                {
+                    if (_sessionHandle == null)
+                    {
+                        uint accessType;
+
+                        // If a custom proxy is specified and it is really the system web proxy
+                        // (initial WebRequest.DefaultWebProxy) then we need to update the settings
+                        // since that object is only a sentinel.
+                        if (state.WindowsProxyUsePolicy == WindowsProxyUsePolicy.UseCustomProxy)
+                        {
+                            Debug.Assert(state.Proxy != null);
+                            try
+                            {
+                                state.Proxy.GetProxy(state.RequestMessage.RequestUri);
+                            }
+                            catch (PlatformNotSupportedException)
+                            {
+                                // This is the system web proxy.
+                                state.WindowsProxyUsePolicy = WindowsProxyUsePolicy.UseWinInetProxy;
+                                state.Proxy = null;
+                            }
+                        }
+
+                        if (state.WindowsProxyUsePolicy == WindowsProxyUsePolicy.DoNotUseProxy ||
+                            state.WindowsProxyUsePolicy == WindowsProxyUsePolicy.UseCustomProxy)
+                        {
+                            // Either no proxy at all or a custom IWebProxy proxy is specified.
+                            // For a custom IWebProxy, we'll need to calculate and set the proxy
+                            // on a per request handle basis using the request Uri.  For now,
+                            // we set the session handle to have no proxy.
+                            accessType = Interop.WinHttp.WINHTTP_ACCESS_TYPE_NO_PROXY;
+                        }
+                        else if (state.WindowsProxyUsePolicy == WindowsProxyUsePolicy.UseWinHttpProxy)
+                        {
+                            // Use WinHTTP per-machine proxy settings which are set using the "netsh winhttp" command.
+                            accessType = Interop.WinHttp.WINHTTP_ACCESS_TYPE_DEFAULT_PROXY;
+                        }
+                        else
+                        {
+                            // Use WinInet per-user proxy settings.
+                            accessType = Interop.WinHttp.WINHTTP_ACCESS_TYPE_AUTOMATIC_PROXY;
+                        }
+
+                        _sessionHandle = Interop.WinHttp.WinHttpOpen(
+                            IntPtr.Zero,
+                            accessType,
+                            Interop.WinHttp.WINHTTP_NO_PROXY_NAME,
+                            Interop.WinHttp.WINHTTP_NO_PROXY_BYPASS,
+                            0);
+                        if (!_sessionHandle.IsInvalid)
+                        {
+                            _sessionHandle.DangerousAddRef(ref ignore);
```

Comment:
Actually, it looks like we create this, AddRef, then immediately replace the SafeHandle.  How is this ever released? If it's actually never allocated, why would we call AddRef?  And why automatically AddRef the second created handle?